src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/DescriptorRegistry.cs

```diff
@@ -1,18 +1,30 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Collections.Concurrent;
 using Google.Protobuf.Reflection;
+using Google.Rpc;
 
 namespace Microsoft.AspNetCore.Grpc.JsonTranscoding.Internal;
 
 internal sealed class DescriptorRegistry
 {
+    private readonly object _lock = new object();
     private readonly HashSet<FileDescriptor> _fileDescriptors = new HashSet<FileDescriptor>();
-    private readonly HashSet<EnumDescriptor> _enumDescriptors = new HashSet<EnumDescriptor>();
+    private readonly ConcurrentDictionary<Type, DescriptorBase> _typeDescriptorMap = new ConcurrentDictionary<Type, DescriptorBase>();
+
+    public DescriptorRegistry()
+    {
+        // For Grpc.Rpc.Status, which is used to send error responses.
+        AddFileDescriptorsRecursive(StatusReflection.Descriptor);
+    }
 
     public void RegisterFileDescriptor(FileDescriptor fileDescriptor)
     {
-        AddFileDescriptorsRecursive(fileDescriptor);
+        lock (_lock)
```

Comment:
I feel like I'm missing something obvious, but why lock rather than (e.g.) `ConcurrentDictionary.TryAdd`?

Reply:
I think maybe it's to guard the update of `_fileDescriptors`, but that won't actually make updates atomic, will it?  You'd have to lock the read as well?

Reply:
It is to prevent multiple threads from updating _fileDescriptors simultaneously. _fileDescriptors isn't used for reads, only _typeDescriptorMap, which is why it is a `ConcurrentDictionary` (i.e. safe for reads during updates)

Reply:
Why not use a `ConcurrentBag` or something for `_fileDescriptors`?  I may be overthinking things, but that seems like an unnecessarily long time to hold the lock.  Also, could we eliminate `_fileDescriptors` entirely by just kicking out if we see a collision in `_typeDescriptorMap`?

Reply:
`ConcurrentBag` allows duplicates.

Could use `ConcurrentDictionary<FileDescriptor, byte>`, and ignore the dictionary value.